# Hypothesis Testing
## Parametric Tests for Continuous Data (Investigating Means)
### One Sample T-test against Hypothesised Population Mean
$H_0: Sample\:Mean= Hypothesised\:Population\:Mean$<br>
$H_1: Sample\:Mean\neq Hypothesised\:Population\:Mean$<br>
Assumptions: Sample size >30 or when the population SD is known<br>
Produces sample mean, SE, t-statistic, p(observing a value from SND (T) greater than test statistic t) or Pr(T > t) or right-tailed p-value, 95% CI

In [1]:
import os
os.chdir("/Applications/Stata/utilities")
from pystata import config
config.init("se")

import pandas as pd

IBM = '/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/IBM.dta'
LSS = '/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/LSS.dta'
IBM = pd.read_stata(IBM)
LSS = pd.read_stata(LSS)
print(IBM.head())


  ___  ____  ____  ____  ____ ®
 /__    /   ____/   /   ____/      StataNow 18.5
___/   /   /___/   /   /___/       SE—Standard Edition

 Statistics and Data Science       Copyright 1985-2023 StataCorp LLC
                                   StataCorp
                                   4905 Lakeway Drive
                                   College Station, Texas 77845 USA
                                   800-782-8272        https://www.stata.com
                                   979-696-4600        service@stata.com

Stata license: Unlimited-user network, expiring  9 Sep 2025
Serial number: 501809305305
  Licensed to: Mujie
               

Notes:
      1. Unicode is supported; see help unicode_advice.
      2. Maximum number of variables is set to 5,000 but can be increased;
          see help set_maxvar.
   Patient_ID  Gender  Age  Ageofonset  Dysphagia_BL  Walkingaid_BL  \
0           1       1   54          46             0              0   
1           2       1   83          74

In [2]:
%%stata -d IBM

ttest IBMFRS_baseline == 25


. 
. ttest IBMFRS_baseline == 25

One-sample t test
------------------------------------------------------------------------------
Variable |     Obs        Mean    Std. err.   Std. dev.   [95% conf. interval]
---------+--------------------------------------------------------------------
IBMFRS~e |      30        24.8    1.295793    7.097353     22.1498     27.4502
------------------------------------------------------------------------------
    mean = mean(IBMFRS_baseline)                                  t =  -0.1543
H0: mean = 25                                    Degrees of freedom =       29

    Ha: mean < 25               Ha: mean != 25                 Ha: mean > 25
 Pr(T < t) = 0.4392         Pr(|T| > |t|) = 0.8784          Pr(T > t) = 0.5608

. 


Pr(T > t) is the likelihood of obtaining a test mean as more extreme than t, assuming the null hypothesis is true;<br>
This is also known as the p-value, or probability of incorrectly rejecting the null hypothesis were it true

Pr(T < t) is the likelihood of obtaining a test mean less extreme than t (or the AUC to the left of the t-statistic), assuming the null hypothesis is true
Which is equivalent to 1 - Pr(T > t)

Pr(|T| > |t|) is the two-tailed probability of the test mean being as extreme as the test statistic (usually *2 of right-tailed p-value)<br>
Pr(T > t) is equal Pr(T < -t) of the opposite tail due to the symmetry of the distribution itself

> N.B. The One Sample z-test is a special case in one sample z-tests, whereby the sample means follow the standard normal distribution, a special type of normal distribution whereby population mean = 0 and SD = 1

### Two Sample T-test
$H_0: Mean\:1 = Mean\:2$<br>
$H_1: Mean\:1\neq Mean\:2$<br>
Assumptions: Data follows a normal distribution<br>
Produces means of each group, Δmeans, SE, t-statistic, p(observing a value from SND (T) greater than test statistic t) or Pr(T > t) or right-tailed p-value, 95% CI

In [3]:
%%stata

ttest IBMFRS_1year, by(Gender)


. 
. ttest IBMFRS_1year, by(Gender)

Two-sample t test with equal variances
------------------------------------------------------------------------------
   Group |     Obs        Mean    Std. err.   Std. dev.   [95% conf. interval]
---------+--------------------------------------------------------------------
       0 |       6    18.16667    2.315407    5.671567    12.21472    24.11861
       1 |      24    23.95833    1.570192     7.69234    20.71014    27.20652
---------+--------------------------------------------------------------------
Combined |      30        22.8    1.390774    7.617584    19.95555    25.64445
---------+--------------------------------------------------------------------
    diff |           -5.791667    3.364945               -12.68444    1.101111
------------------------------------------------------------------------------
    diff = mean(0) - mean(1)                                      t =  -1.7212
H0: diff = 0                                     Degre

### Paired Samples T-test
$H_0: Mean_{t1} = Mean_{t2}$<br>
$H_1: Mean_{t1}\neq Mean_{t2}$<br>

or

$H_0: \Delta Mean = 0$<br>
$H_1: \Delta Mean\neq 0$

In [4]:
%%stata

ttest IBMFRS_baseline == IBMFRS_1year

gen IBMFRS_diff = IBMFRS_1year - IBMFRS_baseline
ttest IBMFRS_diff == 0


. 
. ttest IBMFRS_baseline == IBMFRS_1year

Paired t test
------------------------------------------------------------------------------
Variable |     Obs        Mean    Std. err.   Std. dev.   [95% conf. interval]
---------+--------------------------------------------------------------------
IBMFRS~e |      30        24.8    1.295793    7.097353     22.1498     27.4502
IBMFRS~r |      30        22.8    1.390774    7.617584    19.95555    25.64445
---------+--------------------------------------------------------------------
    diff |      30           2    .6643638    3.638871    .6412234    3.358777
------------------------------------------------------------------------------
     mean(diff) = mean(IBMFRS_baseline - IBMFRS_1year)            t =   3.0104
 H0: mean(diff) = 0                              Degrees of freedom =       29

 Ha: mean(diff) < 0           Ha: mean(diff) != 0           Ha: mean(diff) > 0
 Pr(T < t) = 0.9973         Pr(|T| > |t|) = 0.0054          Pr(T > t) =

## Non-Parametric Tests for Continuous Data
Non-parametric Tests are performed on ranked data rather than data themselves e.g. 43 68 112 452 will be coded as 1 2 3 4
> Instead of the t-distribution, non-parametric tests rely on the standard normal distribution (or z-distribution), since the ranking and resampling data leads to approximation of SND due to the Central Limit Theorem (CLM)

### Wilcoxon Rank Sum or Mann Whitney U Test

$H_0: Median\:1 = Median\:2$<br>
$H_1: Median\:1\neq Median\:2$<br>

Produces rank sum values, expected rank sum values, variance adjustment (to account for tied ranks), z-statistic, P-value or Pr > |z| or the chance of observing z-value as high as this under the $H_0$

In [5]:
print(LSS.head())

  PatientID  Operation  Gender  EQ5DPre  EQ5D24mts
0      H001          1       1    0.088      0.516
1      H002          1       0   -0.536        NaN
2      H005          1       0    0.159      0.691
3      H009          1       0    0.587      0.691
4       H12          0       0   -0.016      0.516


In [6]:
%%stata 

use "/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/LSS.dta", clear

ranksum EQ5D24mts, by(Operation)


. 
. use "/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/LSS.dta", clear

. 
. ranksum EQ5D24mts, by(Operation)

Two-sample Wilcoxon rank-sum (Mann–Whitney) test

   Operation |      Obs    Rank sum    Expected
-------------+---------------------------------
           0 |       24         517         516
           1 |       18         386         387
-------------+---------------------------------
    Combined |       42         903         903

Unadjusted variance     1548.00
Adjustment for ties       -8.78
                     ----------
Adjusted variance       1539.22

H0: EQ5D24~s(Operat~n==0) = EQ5D24~s(Operat~n==1)
         z =  0.025
Prob > |z| = 0.9797
Exact prob = 0.9849

. 


### Paired Samples T-test (Wilcoxon Signed Rank)

$H_0: Median_{t1}= Median_{t2}$<br>
$H_1: Median_{t1}\neq Median_{t2}$

or

$H_0: Median\:of\:\Delta= 0$<br>
$H_1: Median\:of\:\Delta \neq 0$

In [7]:
%%stata

signrank EQ5DPre = EQ5D24mts

gen EQ5D_diff = EQ5D24mts - EQ5DPre
signrank EQ5D_diff = 0


. 
. signrank EQ5DPre = EQ5D24mts

Wilcoxon signed-rank test

        Sign |      Obs   Sum ranks    Expected
-------------+---------------------------------
    Positive |       10         157         451
    Negative |       31         745         451
        Zero |        1           1           1
-------------+---------------------------------
         All |       42         903         903

Unadjusted variance     6396.25
Adjustment for ties       -1.88
Adjustment for zeros      -0.25
                     ----------
Adjusted variance       6394.12

H0: EQ5DPre = EQ5D24mts
         z = -3.677
Prob > |z| = 0.0002
Exact prob = 0.0001

. 
. gen EQ5D_diff = EQ5D24mts - EQ5DPre
(5 missing values generated)

. signrank EQ5D_diff = 0

Wilcoxon signed-rank test

        Sign |      Obs   Sum ranks    Expected
-------------+---------------------------------
    Positive |       31         746         451
    Negative |       10         156         451
        Zero |        1           1   

### 95% Confidence Interval
For centile difference as opposed to "ci" for mean difference

In [8]:
%%stata

centile EQ5D_diff, centile(50)


. 
. centile EQ5D_diff, centile(50)

                                                          Binom. interp.   
    Variable |       Obs  Percentile    Centile        [95% conf. interval]
-------------+-------------------------------------------------------------
   EQ5D_diff |        42         50       .2745            .104    .3526933

. 


If 95% CI is narrow and exclude 0, $H_0$ can be rejected

### Other Tests:
Means Test of Continuous Data between <u>>2 Groups</u>
- 1-way ANOVA (P)
- Kruskal-Wallis ANOVA (NP)

Means Test of Paired Continuous Data between <u>> 2 Groups</i>
- Repeated Measures ANOVA (P)

____________________________________________________________________________________________________
## Parametric Tests for Categorical Data (Investigating Proportions)
### One Sample Z-test against Hypothesised Population Proportion
$H_0: Sample\: Proportion^* = Hypothesised\: Population\: Proportion$<br>
$H_1: Sample\: Proportion \neq Hypothesised\: Population\: Proportion$<br>
*where the outcomes are binary; Bernoulli trials
> The binomial distribution approximates the normal distibution when the sample size is large enough

Produces sample Pr(Success), SE, 95% CI, z-statistic, p-values for one-tailed and two-tailed tests:

In [9]:
%%stata

prtest Gender == 0.5 // Where the first argument is a binary variable that indicates success (1) or failure (0)


. 
. prtest Gender == 0.5 // Where the first argument is a binary variable that in
> dicates success (1) or failure (0)

One-sample test of proportion                   Number of obs      =        47

------------------------------------------------------------------------------
    Variable |       Mean   Std. err.                     [95% conf. interval]
-------------+----------------------------------------------------------------
      Gender |   .3829787   .0709069                      .2440038    .5219537
------------------------------------------------------------------------------
    p = proportion(Gender)                                        z =  -1.6045
H0: p = 0.5

     Ha: p < 0.5                 Ha: p != 0.5                   Ha: p > 0.5
 Pr(Z < z) = 0.0543         Pr(|Z| > |z|) = 0.1086          Pr(Z > z) = 0.9457

. 


Pr(Z < z) = 0.0543 (one-tailed test for p < 0.5) not below alpha, unable to reject the null hypothesis<br>
Pr(|Z| > |z|) = 0.1086 (two-tailed test)<br>
Pr(Z > z) = 0.9457 (one-tailed test for p > 0.5)

Since the 95% CI includes the hypothesised population mean, we are unable reject the null hypothesis

### Chi-Squared Test for Two or More Proportions
$H_0: Proportion\:1 = Proportion\:2$<br>
$H_1: Proportion\:1\neq Proportion\:2$<br>
Produces proportions of each group, Δproportions, SE, z-statistic, p(observing a value from SND (Z) greater than test statistic z) or Pr(Z > z) or right-tailed p-value, 95% CI
> N.B If any of the expected frequencies in the contingency table are <5, we cannot assume the data follows a chi-square distribution; Non-parametric tests like the Fisher's Exact Test is more appropriate

In [10]:
%%stata

tab Operation Gender, expected
prtest Operation, by(Gender)
prtesti 29 12 18 9, count // The immediate test whereby proportions are known without providing data
// total_0 success_0 total_1 success_1, where 0 is female and 1 is male

tab Operation Gender, chi2


. 
. tab Operation Gender, expected

+--------------------+
| Key                |
|--------------------|
|     frequency      |
| expected frequency |
+--------------------+

           |        Gender
 Operation |         0          1 |     Total
-----------+----------------------+----------
         0 |        17          9 |        26 
           |      16.0       10.0 |      26.0 
-----------+----------------------+----------
         1 |        12          9 |        21 
           |      13.0        8.0 |      21.0 
-----------+----------------------+----------
     Total |        29         18 |        47 
           |      29.0       18.0 |      47.0 

. prtest Operation, by(Gender)

Two-sample test of proportions                     0: Number of obs =       29
                                                   1: Number of obs =       18
------------------------------------------------------------------------------
       Group |       Mean   Std. err.      z    P>|z|     [9

Pr(Z > z) is used to determine the likelihood of obtaining a test proportion more extreme than z, assuming the null hypothesis is true;<br>
This is also known as the p-value, or probability of incorrectly rejecting the null hypothesis were it true

> For more than 2 groups/samples, an overall p-value will be produced

## Non-Parametric Tests for Categorical Data
### Fischer's Exact Test

Preferred when the sample size or expected frequencies are not sufficiently large
$H_0: \text{Baby's weight has no association with the mother's cigarette numbers}$
$H_1: \text{Baby's weight is associated with the mother's cigarette numbers}$

In [11]:
%%stata

use "/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/babies.dta", clear

// Dichotomise continuous variables into binary groups
gen babylow =.
replace babylow = 0 if bwt <2500
replace babylow = 1 if bwt >= 2500 & bwt !=.

recode cigs (0=0)(1/9=1)(10/19=2)(20/max=3), gen(cigs_cat)
tab babylow cigs_cat, expected

// All expected frequencies are above 5, but let's say they weren't so we can use Fisher's Exact Test
tab babylow cigs_cat, exact


. 
. use "/Users/mujiechen/Jupyter-Notebook/STATA/Datasets/babies.dta", clear

. 
. // Dichotomise continuous variables into binary groups
. gen babylow =.
(997 missing values generated)

. replace babylow = 0 if bwt <2500
(59 real changes made)

. replace babylow = 1 if bwt >= 2500 & bwt !=.
(938 real changes made)

. 
. recode cigs (0=0)(1/9=1)(10/19=2)(20/max=3), gen(cigs_cat)
(329 differences between cigs and cigs_cat)

. tab babylow cigs_cat, expected

+--------------------+
| Key                |
|--------------------|
|     frequency      |
| expected frequency |
+--------------------+

           |               RECODE of cigs
   babylow |         0          1          2          3 |     Total
-----------+--------------------------------------------+----------
         0 |        26          4         19         10 |        59 
           |      38.8        6.6        8.2        5.4 |      59.0 
-----------+--------------------------------------------+----------
         1 |  

A Fisher's exact of 0.000 is sufficient to reject the null hypothesis that there is no difference in birth weights between mothers' cigarette numbers

If proportions are significantly different then the variables are not independent; Proportions will be the same if the variables are independent<br>
For example, the proportion chocolates over strawberries will be equal between boys and girls if the preference for chocolates is independent of the gender; A difference in proportion indicates that there is some relation between flavour and gender

### Multiple Proportions with Multiple Categories
For example, in Adolescents, Adults, and Elderly, compare the proportions of wins, losses, and draws<br>
Use Chi-squared Test